File: python-sdk/src/ace_sdk/pke/elgamal.py
# Copyright (c) Aptos Labs
# SPDX-License-Identifier: Apache-2.0
"""Mirrors src/pke/elgamal.ts: exponential ElGamal over pke/group.py's ristretto255."""

from __future__ import annotations

from ace_sdk.bcs import Deserializer, Serializer
from ace_sdk.pke.group import Element, Scalar


class Ciphertext:
    def __init__(self, c0: Element, c1: Element) -> None:
        self.c0 = c0
        self.c1 = c1

    @staticmethod
    def decode(deserializer: Deserializer) -> "Ciphertext":
        c0 = Element.decode(deserializer)
        c1 = Element.decode(deserializer)
        return Ciphertext(c0, c1)

    def encode(self, serializer: Serializer) -> None:
        self.c0.encode(serializer)
        self.c1.encode(serializer)

    def to_bytes(self) -> bytes:
        serializer = Serializer()
        self.encode(serializer)
        return serializer.to_bytes()

    def add(self, other: "Ciphertext") -> "Ciphertext":
        return Ciphertext(self.c0.add(other.c0), self.c1.add(other.c1))

    def scale(self, scalar: Scalar) -> "Ciphertext":
        return Ciphertext(self.c0.scale(scalar), self.c1.scale(scalar))


class DecKey:
    def __init__(self, enc_base: Element, private_scalar: Scalar) -> None:
        self.enc_base = enc_base
        self.private_scalar = private_scalar

    @staticmethod
    def decode(deserializer: Deserializer) -> "DecKey":
        enc_base = Element.decode(deserializer)
        private_scalar = Scalar.decode(deserializer)
        return DecKey(enc_base, private_scalar)

    def encode(self, serializer: Serializer) -> None:
        self.enc_base.encode(serializer)
        self.private_scalar.encode(serializer)


class EncKey:
    def __init__(self, enc_base: Element, public_point: Element) -> None:
        self.enc_base = enc_base
        self.public_point = public_point

    @staticmethod
    def decode(deserializer: Deserializer) -> "EncKey":
        enc_base = Element.decode(deserializer)
        public_point = Element.decode(deserializer)
        return EncKey(enc_base, public_point)

    def encode(self, serializer: Serializer) -> None:
        self.enc_base.encode(serializer)
        self.public_point.encode(serializer)

    def to_bytes(self) -> bytes:
        serializer = Serializer()
        self.encode(serializer)
        return serializer.to_bytes()


def enc(ek: EncKey, randomizer: Scalar, ptxt: Element) -> Ciphertext:
    return Ciphertext(ek.enc_base.scale(randomizer), ptxt.add(ek.public_point.scale(randomizer)))


def dec(dk: DecKey, ciph: Ciphertext) -> Element:
    unblinder = ciph.c0.scale(dk.private_scalar)
    return ciph.c1.sub(unblinder)


def multi_exp(ciphs: list[Ciphertext], scalars: list[Scalar]) -> Ciphertext:
    acc = Ciphertext(Element.group_identity(), Element.group_identity())
    for ciph, scalar in zip(ciphs, scalars):
        acc = acc.add(ciph.scale(scalar))
    return acc

File: rust-sdk/src/pke/elgamal.rs
// Copyright (c) Aptos Labs
// SPDX-License-Identifier: Apache-2.0

//! Mirrors `ts-sdk/src/pke/elgamal.ts`: plain ElGamal over Ristretto255.

use super::group::{Element, Scalar};
use crate::error::Result;
use crate::wire::{Deserializer, Serializer};

/// BCS: `Element(c0) ++ Element(c1)`.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Ciphertext {
    pub c0: Element,
    pub c1: Element,
}

impl Ciphertext {
    pub fn new(c0: Element, c1: Element) -> Self {
        Self { c0, c1 }
    }

    pub fn decode(d: &mut Deserializer<'_>) -> Result<Self> {
        let c0 = Element::decode(d)?;
        let c1 = Element::decode(d)?;
        Ok(Self { c0, c1 })
    }

    pub fn encode(&self, s: &mut Serializer) {
        self.c0.encode(s);
        self.c1.encode(s);
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut s = Serializer::new();
        self.encode(&mut s);
        s.into_bytes()
    }

    pub fn add(&self, other: &Ciphertext) -> Result<Ciphertext> {
        Ok(Ciphertext::new(
            self.c0.add(&other.c0)?,
            self.c1.add(&other.c1)?,
        ))
    }

    pub fn scale(&self, scalar: &Scalar) -> Result<Ciphertext> {
        Ok(Ciphertext::new(
            self.c0.scale(scalar)?,
            self.c1.scale(scalar)?,
        ))
    }
}

/// BCS: `Element(enc_base) ++ Scalar(private_scalar)`.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct DecKey {
    pub enc_base: Element,
    pub private_scalar: Scalar,
}

impl DecKey {
    pub fn new(enc_base: Element, private_scalar: Scalar) -> Self {
        Self {
            enc_base,
            private_scalar,
        }
    }

    pub fn decode(d: &mut Deserializer<'_>) -> Result<Self> {
        let enc_base = Element::decode(d)?;
        let private_scalar = Scalar::decode(d)?;
        Ok(Self {
            enc_base,
            private_scalar,
        })
    }

    pub fn encode(&self, s: &mut Serializer) {
        self.enc_base.encode(s);
        self.private_scalar.encode(s);
    }
}

/// BCS: `Element(enc_base) ++ Element(public_point)`.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct EncKey {
    pub enc_base: Element,
    pub public_point: Element,
}

impl EncKey {
    pub fn new(enc_base: Element, public_point: Element) -> Self {
        Self {
            enc_base,
            public_point,
        }
    }

    pub fn decode(d: &mut Deserializer<'_>) -> Result<Self> {
        let enc_base = Element::decode(d)?;
        let public_point = Element::decode(d)?;
        Ok(Self {
            enc_base,
            public_point,
        })
    }

    pub fn encode(&self, s: &mut Serializer) {
        self.enc_base.encode(s);
        self.public_point.encode(s);
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let mut s = Serializer::new();
        self.encode(&mut s);
        s.into_bytes()
    }
}

/// `(enc_base * r, ptxt + public_point * r)`.
pub fn enc(ek: &EncKey, randomizer: &Scalar, ptxt: &Element) -> Result<Ciphertext> {
    Ok(Ciphertext::new(
        ek.enc_base.scale(randomizer)?,
        ptxt.add(&ek.public_point.scale(randomizer)?)?,
    ))
}

/// `c1 - c0 * sk`.
pub fn dec(dk: &DecKey, ciph: &Ciphertext) -> Result<Element> {
    let unblinder = ciph.c0.scale(&dk.private_scalar)?;
    ciph.c1.sub(&unblinder)
}

/// `sum_i ciphs[i] * scalars[i]` (component-wise).
pub fn multi_exp(ciphs: &[Ciphertext], scalars: &[Scalar]) -> Result<Ciphertext> {
    let mut acc = Ciphertext::new(Element::group_identity(), Element::group_identity());
    for (c, s) in ciphs.iter().zip(scalars.iter()) {
        acc = acc.add(&c.scale(s)?)?;
    }
    Ok(acc)
}
